scripts/utility.py

```diff
@@ -0,0 +1,49 @@
+import threading
+from functools import wraps
+from typing import Callable, Optional, TypeVar, overload
+
+from typing_extensions import ParamSpec
+
+try:
+    import thread
+except ImportError:
+    import _thread as thread
+
+P = ParamSpec("P")
+T = TypeVar("T")
+
+
+def _interrupt_handler() -> None:
+    thread.interrupt_main()
+
+
+@overload
+def timeout(func: Callable[P, T], time: int, silent: bool = False) -> Callable[P, T]:
+    ...
+
+
+@overload
+def timeout(func: Callable[P, T], time: int, silent: bool = True) -> Callable[P, Optional[T]]:
+    ...
+
+
+def timeout(func: Callable[P, T], time: int, silent: bool = False) -> Callable[P, Optional[T]]:
+    @wraps(func)
+    def wrapper(*args: P.args, **kwargs: P.kwargs) -> Optional[T]:
+        # register interrupt handler
+        timer = threading.Timer(time, _interrupt_handler)
+        timer.start()
+
+        try:
+            result = func(*args, **kwargs)
+        except KeyboardInterrupt as err:
+            if silent:
+                return None
+            else:
+                raise TimeoutError from err
+        finally:
+            timer.cancel()
+            timer.join()
```

Comment:
Just to clarify: The timer needs to get cancelled if the `KeyboardInterrupt` is actually thrown by user action and not by the `_interrupt_handler`. Thus the timer may not have been finished.

Also, what is the purpose of `.join`? Shouldn't `.cancel` do everything that's needed as described in the documentation?

Reply:
The timer is canceled when a result comes in or a `KeybordInterupt` is raised, either by user action or by `thread.interrupt_main()`.

The join is just for safety reasons. I couldn't find a reference in the documentation that the thread is joined when using calling `cancel`. Could you give me a link?

Reply:
The only documentation I could find was [this](https://docs.python.org/3/library/threading.html#threading.Timer.cancel).

> Stop the timer, and cancel the execution of the timer’s action. This will only work if the timer is still in its waiting stage.

I don't know if it will be joined silently or not. I just assumed that this method takes care of everything, since stopping the timer is all we require "directly". [This](https://github.com/python/cpython/blob/b6839942a8906fccdd64e749abeefe8a61ce7e03/Lib/threading.py#L1333) is the line in the cpython source code but with my knowledge it does not help.

Reply:
I will remove the `join`